### PR TITLE
Low-frame rate Speech Codec Implementation

### PR DESCRIPTION
**Pull request by Edresson**
# What does this PR do ?

It adds the implementation of our latest 21Hz Audio Codec called "Low-frame rate Speech Codec"

**Collection**:  [TTS]

# Changelog 
- Add a WavLM-based discriminator which improves the performance of the codec in low bitrate scenario
- Add config files to training 22khz NeMo Speech Codec model.

# GitHub Actions CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

The GitHub Actions CI will run automatically when the "Run CICD" label is added to the PR.
To re-run CI remove and add the label again.
To run CI on an untrusted fork, a NeMo user with write access must first click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [x] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [x] Did you write any new necessary tests?
- [x] Did you add or update any necessary documentation?
- [x] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [x] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [ ] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

### CODE REVIEW

File: tutorials/tts/Audio_Codec_Training.ipynb [hunk 1]
```diff
@@ -130,26 +133,29 @@
         "*   **audio_codec_*.yaml**: Audio codec configurations optimized for various sampling rates.\n",
         "*   **mel_codec_*.yaml**: A mel-spectrogram based codec designed to maximize the performance of speech synthesis models.\n",
         "*   **encodec_*.yaml**: A reproduction of the original [EnCodec](https://arxiv.org/abs/2210.13438) model setup.\n",
+        "*   **audio_codec_low_frame-rate_22050.yaml**: The Low Frame-rate Speech Codec: an audio codec that achieves high-quality audio compression with a 1.89 kbps bitrate and 21.5 frames per second.\n",
```
Comment (rlangman) [inline tutorials/tts/Audio_Codec_Training.ipynb:136]: Nitpick: I would favor underscore over hyphen in a file name, like "frame_rate".

**Comment by prassanna-ravishankar** [comment]
Where is the model checkpoint?

**Comment by rlangman** [comment]
> Where is the model checkpoint?

The checkpoint is currently undergoing legal review. We will provide an ETA when we are closer to finishing our publication review process.

File: examples/tts/conf/audio_codec/low_frame-rate_speech_codec_22050.yaml [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+  steps_per_epoch: ${weighted_sampling_steps_per_epoch}
+
+  sample_rate: ${sample_rate}
+  samples_per_frame: ${samples_per_frame}
+
+  mel_loss_l1_scale: 10.0
+  mel_loss_l2_scale: 0.0
+  stft_loss_scale: 10.0
+  time_domain_loss_scale: 0.0
+  commit_loss_scale: 0.0
```
[the remark this answers is not in the document]
Comment (Edresson) [reply examples/tts/conf/audio_codec/low_frame-rate_speech_codec_22050.yaml]: Good catch I updated the config to match the right loss weights.

File: examples/tts/conf/audio_codec/low_frame-rate_speech_codec_22050.yaml [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+  # Probability of updating the discriminator during each training step
+  # For example, update the discriminator 1/2 times (1 update for every 2 batches)
+  disc_updates_per_period: 1
+  disc_update_period: 2
+
+  # All resolutions for mel reconstruction loss, ordered [num_fft, hop_length, window_length]
+  loss_resolutions: [
+    [32, 8, 32], [64, 16, 64], [128, 32, 128], [256, 64, 256], [512, 128, 512], [1024, 256, 1024]
+  ]
+  mel_loss_dims: [5, 10, 20, 40, 80, 160]
```
[the remark this answers is not in the document]
Comment (Edresson) [reply examples/tts/conf/audio_codec/low_frame-rate_speech_codec_22050.yaml]: Good catch, I updated the config.

File: examples/tts/conf/audio_codec/low_frame-rate_speech_codec_22050.yaml [hunk 4]
```diff
@@ -0,0 +1,199 @@
...
+  feature_loss_type: absolute
+
+  train_ds:
+    dataset:
+      _target_: nemo.collections.tts.data.vocoder_dataset.VocoderDataset
+      dataset_meta: ${train_ds_meta}
+      weighted_sampling_steps_per_epoch: ${weighted_sampling_steps_per_epoch}
+      sample_rate: ${sample_rate}
+      n_samples: ${train_n_samples}
+      min_duration: 0.4 # seconds
```
[the remark this answers is not in the document]
Comment (Edresson) [reply examples/tts/conf/audio_codec/low_frame-rate_speech_codec_22050.yaml]: Yes, anything shorter is automatically padded to the appropriate training length. I decided to use 0.4 to not filter too many audios and make internal codec comparisons unfair.

File: tutorials/tts/Audio_Codec_Training.ipynb [hunk 5]
```diff
@@ -130,26 +133,29 @@
...
+      "source": [
+        "from omegaconf import OmegaConf"
+      ]
     },
     {
       "cell_type": "code",
+      "execution_count": null,
+      "metadata": {
+        "id": "iCPJFKg63Dsv"
+      },
```
[the remark this answers is not in the document]
Comment (Edresson) [reply tutorials/tts/Audio_Codec_Training.ipynb:157]: I think it is possible for me to edit it using a code editor that does not support notebook editing and then edit directly the metadata. But then, if I run to test it, it will automatically update the metadata. Then I'm not sure how to do so.

File: nemo/collections/tts/modules/audio_codec_modules.py [hunk 6]
```diff
@@ -56,6 +58,87 @@ def get_up_sample_padding(kernel_size: int, stride: int) -> Tuple[int, int]:
...
+        input_sr=22050,
+        slm_hidden=768,
+        slm_layers=13,
+        initial_channel=64,
+        use_spectral_norm=False,
+    ):
+        super().__init__()
+
+        self.slm_model = SSLModel(slm_model_name)
+        self.slm_model.ssl_model.feature_extractor._requires_grad = False
```
[the remark this answers is not in the document]
Comment (Edresson) [reply nemo/collections/tts/modules/audio_codec_modules.py:98]: Good catch I think `self.slm_model.ssl_model.feature_extractor._requires_grad = False ` and `.freeze()` are doing the same thing in the end, but `.freeze()` is not sensitive to the `.train()` `.eval()` calls so I will remove `self.slm_model.ssl_model.feature_extractor._requires_grad = False `.

File: nemo/collections/tts/modules/audio_codec_modules.py [hunk 7]
```diff
@@ -56,6 +58,87 @@ def get_up_sample_padding(kernel_size: int, stride: int) -> Tuple[int, int]:
...
+            x = l(x)
+            x = F.leaky_relu(x, 0.1)
+            fmap.append(x.unsqueeze(-1))
+
+        x = self.conv_post(x)
+        x = torch.flatten(x, 1, -1)
+
+        return x, fmap
+
+    def forward(self, audio_real, audio_gen):
```
[the remark this answers is not in the document]
Comment (Edresson) [reply nemo/collections/tts/modules/audio_codec_modules.py:149]: Done

File: tutorials/tts/Audio_Codec_Training.ipynb [hunk 8]
```diff
@@ -746,44 +747,44 @@
...
     {
       "cell_type": "markdown",
+      "metadata": {
+        "id": "Rvu4w2x_3RSY"
+      },
       "source": [
         "1.   [EnCodec](https://arxiv.org/abs/2210.13438)\n",
         "2.   [Finite Scalar Quantization (FSQ)](https://arxiv.org/abs/2309.15505)\n",
         "3.   [HiFi-GAN](https://arxiv.org/abs/2010.05646)\n",
         "4.   [SEANet](https://arxiv.org/abs/2009.02095)"
```
Comment (rlangman) [inline tutorials/tts/Audio_Codec_Training.ipynb:786]: Should we add our 2 papers to the references?

**Review by rlangman — APPROVED** [review]

File: tutorials/tts/Audio_Codec_Training.ipynb (hunk 8 above)
Comment (Edresson) [reply tutorials/tts/Audio_Codec_Training.ipynb]: Done

**Review by rlangman — APPROVED** [review]

**Review by rlangman — APPROVED** [review]

**Review by rlangman — APPROVED** [review]
